allennlp/training/optimizers.py

```diff
@@ -196,6 +205,159 @@ def default(model_parameters: List) -> "Optimizer":
+        should follow the same format as there, except an additional 'name' argument should be provided
+        to match this group to its own optimizer. Optimizer options can also be set for this group which
+        will override the default options.
+    """
+
+    def __init__(
+        self,
+        model_parameters: List[Tuple[str, torch.nn.Parameter]],
+        optimizers: List[Dict[str, Any]],
+        parameter_groups: List[Tuple[List[str], Dict[str, Any]]],
+    ):
+        self.model_parameters = model_parameters
+        self.optimizers = optimizers
+        self.parameter_groups = parameter_groups
+
+        self._grouped_optimizers: Dict[str, Optimizer] = {}
+
+        optimizer_names = [optimizer["name"] for optimizer in self.optimizers]
+        assert "default" in optimizer_names, (
+           "No optimizer was provided for the 'default' group."
+           " Please provide a 'name': 'default' parameter for one of the `optimizers`."
+           )
+
+        parameter_groups = make_parameter_groups(self.model_parameters, self.parameter_groups)
+        # Manually set optimizer kwargs for the default group, so certain schedulers are able
+        # to know the options for this group.
+        if len(list(parameter_groups[-1].keys())) == 1:
+            for optimizer in self.optimizers:
+                if optimizer["name"] == "default":
+                    parameter_groups[-1].update(optimizer)
+
+        optimizer_groups = self._populate_optimizer_groups(parameter_groups)
+        self._set_optimizer_kwargs(optimizer_groups, self.optimizers)
+
+        # Check to see if an optimizer didn't receive any parameters.
+        for optimizer_key, optimizer_parameters in optimizer_groups.items():
+            if len(optimizer_parameters[0]) == 0:
+               raise ValueError(
+                    f"Optimizer '{optimizer_key}' did not receive any parameters."
+                    " If you are using `parameter_groups`, please make sure that the regexes you have provided"
+                    " match the desired model parameters, or that the `name` value of this optimizer "
+                    " matches that of the parameter group you are trying to assign to it."
+                    " Alternatively, you can remove this optimizer from the provided `optimizers`"
+                    " if it is not relevant to a particular parameter group."
+                    )
+
+        # We have already created our parameter groups so we set `parameter_groups` to False, which is a no-op in `make_parameter_groups`.
+        # This is so that we don't have to change any of the code when initializing individual optimizers.
+        for optimizer_name, (params, optimizer_kwargs) in optimizer_groups.items():
+            self._grouped_optimizers[optimizer_name] = Optimizer.from_params(model_parameters=params, parameter_groups=False, params=Params(optimizer_kwargs))
+        
+        super().__init__(parameter_groups, optimizer_groups["default"][1])
```

Comment:
This is how the `RegexOptimizer`'s `parameter_groups` are accessed in the `Trainer` where they can be modified. The main `kwargs` (`optimizer_groups["default"][1]`) for the `RegexOptimizer` class are never used but it is a requirement for initializing this class. I think I should add another test for this to be sure.

Reply:
I like your change where the optimizer options are now initialized with an empty `Dict` which makes it less likely for the outer optimizer to do anything unexpected with some placeholder options.